design/healcheck-endpoint.md

```diff
@@ -0,0 +1,47 @@
+# Healthcheck endpoint
+
+* **Status**: Draft #1
+* **JIRA**: [KEYCLOAK-12398](https://issues.redhat.com/browse/KEYCLOAK-12398)
+
+## Motivation
+
+Deploying both Keycloak and RHSSO on top of Kubernetes requires defining both Readiness Check and Liveness Check (see the [manual](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)). Currently for Keycloak we use a very naive implementation, that calls `/auth/realms/master` URL. This approach has many drawbacks, here are a few examples:
+
+* We can not discover database failures
+* We can not discover if certain Identity Provider is up and running
+* Undertow subsystem is down when proceeding with migration
+
+In order to address this problems, we should provide a parametrized Healthcheck endpoint in the Server and use it in our container image as well as in the Operator. 
+
+## Parametrizing the Healthcheck URL
+
+The preferable approach is to hook into Health Endpoint provided by Wildfly (starting from 16) at:
+
+    http://localhost:9990/health
+
+The endpoint is provided according to the [MicroProfile Health](https://microprofile.io/project/eclipse/microprofile-health) specification.
+
+The implementation should take additional parameters into consideration, including:
+
+* `timeout` - the request gets terminated if not performed within a timeout. The implementation needs to take special care of releasing the thread pool (if necessary). Subsequent calls that take longer than expected should not exhaust the thread pools inside Wildfly.
+* `fail-on` - a set of conditions under which the probe fails. The initial implementation should cover:
+  * `db` - failing on database failures
+  * `idp-<prefix>` - failing when an IdP is not accessible
```

Comment:
Maybe we should have an option to fail on all idp's?

Reply:
`<prefix>` doesn't make sense here, should be `<identity provider alias>`

Reply:
> Maybe we should have an option to fail on all idp's?

+1
But I'd keep the option to ping a specific IdP as well.

Reply:
What would the IdP check do? Just verify if the IdP is online? That doesn't tell us much, does it? Let's say an IdP changes API and login stops working for us. A simple online check won't discover this.

Reply:
There are two ways to do that:
- Active - we send a request to an external IDP and see if it's up. Not perfect as you mentioned but it's fairly easy to implement.
- Passive - one of the things mentioned in the [Observability Design](https://github.com/keycloak/keycloak-community/pull/177/files#diff-5c0706d1dc103d26e66e94c428afd346R150) is passive tracking of the number of exceptions from a specific external IdP. We can leverage this mechanism here and implement a switch - if we observe 10 exceptions in a certain amount of time, we assume the external IdP has some problems. Otherwise, we assume it's fine.

I'm leaning towards the passive mode. I know it might be less accurate, but it's actually "free" for us.

Reply:
+1 for the passive mode, I like it! But we could perhaps combine both approaches.

Reply:
Perhaps... But I would need very specific examples here.

Reply:
After having a discussion with @stianst, we decided to remove both External IdPs and User Federation from the Healthcheck API.

Here's the most important argument for doing this:

> Including both checks probably makes the end-user situation worse. If we include an External IdP check in the Health endpoint, Kubernetes will plug all the pods out of the Service (a load balancer) if that IdP is down. Once all pods are removed from the Service, an end-user will see an HTTP 503 Error page generated by the OpenShift router. If however, the External IdP is down and we don't include it in the Health endpoint, a user will probably face an error during log in. But the error page will be displayed by Keycloak and we control what message we send there. I consider this as a slightly better situation, although the end-result is the same - a user can not log in.

Reply:
Please do not resolve this conversation. It should stay here as a reference why we decided to leave the External IdPs and User Federation out.